ringpop.thrift

```diff
@@ -0,0 +1,25 @@
+struct Change {
+    1: required string source;
+    2: required string address;
+    3: required string status;
+    4: required double incarnationNumber;
+}
+
+struct JoinResult {
+    1: required list<Change> changes;
+}
+
+struct PingResult {
+    1: required list<Change> changes;
+}
+
+struct PingReqResult {
+    1: required bool pingStatus;
+    2: required list<Change> changes;
+}
+
+service Ringpop {
+    JoinResult join(1:string app, 2:string source, 3:double incarnationNumber);
+    PingResult ping(1:double checksum, 2:list<Change> changes);
+    PingReqResult pingReq(1:double checksum, 2:list<Change> changes, 3:string target);
+}
```

Comment:
I really like the consolidation of PingResult/PingReqResult into Ack, and it works well for now (I tried it), but I reverted since I felt that it was wishful thinking that the semantics would always be the same.

I also like the consolidation of the join operation into ping. There are some slightly different semantics from the way join and ping work now:
* incarnation number is exchanged during join (and not ping)
* full-sync is performed on join (and not ping)
* app is exchanged during join (and not ping)

The first and third are easily rectified by including them in the ping interface. The second is easy too, but the more I try to bend joining around the ping, the less natural the whole refactor feels.

I'm going to give it a shot though.